### PR TITLE
Update lcobucci/jwt requirement to ^4.1

### PR DESCRIPTION
**Pull request by Slamdunk**
* [x] Blocked by lcobucci/jwt#554
 * [x] Blocked by lcobucci/jwt#555

BC Breaks at type level:

```
[BC] REMOVED: Constant PSR7Sessions\Storageless\Http\SessionMiddleware::ISSUED_AT_CLAIM was removed
[BC] REMOVED: Method PSR7Sessions\Storageless\Http\SessionMiddleware::fromAsymmetricKeyDefaults() was removed
[BC] CHANGED: The parameter $signer of PSR7Sessions\Storageless\Http\SessionMiddleware#__construct() changed from Lcobucci\JWT\Signer to a non-contravariant Lcobucci\JWT\Configuration
[BC] CHANGED: The parameter $signatureKey of PSR7Sessions\Storageless\Http\SessionMiddleware#__construct() changed from string to a non-contravariant Dflydev\FigCookies\SetCookie
[BC] CHANGED: The parameter $verificationKey of PSR7Sessions\Storageless\Http\SessionMiddleware#__construct() changed from string to a non-contravariant int
[BC] CHANGED: The parameter $defaultCookie of PSR7Sessions\Storageless\Http\SessionMiddleware#__construct() changed from Dflydev\FigCookies\SetCookie to a non-contravariant Lcobucci\Clock\Clock
[BC] CHANGED: The parameter $tokenParser of PSR7Sessions\Storageless\Http\SessionMiddleware#__construct() changed from Lcobucci\JWT\Parser to a non-contravariant int
[BC] CHANGED: The parameter $symmetricKey of PSR7Sessions\Storageless\Http\SessionMiddleware::fromSymmetricKeyDefaults() changed from string to a non-contravariant Lcobucci\JWT\Signer\Key
```

### CODE REVIEW

File: src/Storageless/Http/SessionMiddleware.php
Comment (Slamdunk) [inline src/Storageless/Http/SessionMiddleware.php]: I propose to change the signature here to accept any [`Lcobucci\JWT\Signer\Key`](https://github.com/lcobucci/jwt/tree/master/src/Signer/Key)

File: src/Storageless/Http/SessionMiddleware.php
Comment (Slamdunk) [inline src/Storageless/Http/SessionMiddleware.php]: See https://github.com/lcobucci/jwt/issues/554

**Review by Ocramius — CHANGES_REQUESTED** [review]
For the failing assertions, I think we need `roave/infection-static-analysis-plugin`

For the BC breaks, I suppose we need a new major? Is there a way to mitigate them?

**Comment by Slamdunk** [comment]
> For the BC breaks, I suppose we need a new major? Is there a way to mitigate them?

Maybe with `SessionMiddleware::__construct` marked as `@internal`, but not otherwise.

We can manage to handle `$signer`, `$signatureKey` and `$verificationKey` constructor parameters both for `v3` and `v4`, but not the `$tokenParser` one, which is absent in `v4` and would raise an error to anyone updating to an hypothetical `storageless` new MINOR; hence a new MAJOR is needed.

**Comment by Ocramius** [comment]
Marking the CTOR `@internal` would also be a BC break.

I'm OK with making a new major release for this BTW: mine were genuine questions about feasibility :-)

**Comment by Slamdunk** [comment]
Ok. Then, what about https://github.com/psr7-sessions/storageless/pull/146#discussion_r530319831?

File: src/Storageless/Http/SessionMiddleware.php
Comment (Ocramius) [reply src/Storageless/Http/SessionMiddleware.php]: Not sure about this one: I guess it is better to expose `lcobucci/jwt` API here though, instead of requiring a `string`.

Comment (Slamdunk) [reply src/Storageless/Http/SessionMiddleware.php]: So we agree, fine :+1:

**Comment by Ocramius** [comment]
**Important note to self**: change base branch ref before merging

**Comment by Slamdunk** [comment]
It seems the solution is going to be ready soon: https://github.com/lcobucci/jwt/pull/555#issuecomment-741725761

Stay tuned :mag:

**Comment by Ocramius** [comment]
One important test to be tried here:

 1. Generate a session with current stable
 2. Use that session with this branch and verify it is still usable

**Comment by Slamdunk** [comment]
I can assure you that that test is going to **fail**, due to the fact that I pushed the validator I'm going to propose here to require the `Not Before` claim, which `storageless:v7` never set:

https://github.com/psr7-sessions/storageless/blob/20a60d5ef18ca4d895931f79b09eba17463b9ba5/src/Storageless/Http/SessionMiddleware.php#L248-L253

**Comment by Ocramius** [comment]
Eh, good thing I asked then...

Don't think that should be the case for now, unless there are security implications that are clear?

**Comment by Slamdunk** [comment]
We can draw our own `StrictValidAt` validator without mandatory `not-before` claim, if you prefer, but:

> 7.0.0
> This change is a major BC break, since upgrading the library will now lead to active sessions being dropped when deploying an application with this new version.

We already had active session dropping releases, and this PR is already requiring a new MAJOR, so why petting the users when they're already used to be pushed down the chasm of caring when to deploy a session breaking change? :speak_no_evil:

**Comment by Ocramius** [comment]
> so why petting the users when they're already used to be pushed down the chasm of caring when to deploy a session breaking change? :speak_no_evil:

This is a data change: deploying this will not just have an effect on code, but will also cause losing all sessions for active users on deployment, which is not to be underrated.

I'd much rather split it out to a further major, if needed at all.

**Comment by Slamdunk** [comment]
> This is a data change: deploying this will not just have an effect on code, but will also cause losing all sessions for active users on deployment, which is not to be underrated.

I'm sorry I don't understand: `v7` required losing all sessions for active users on deployment, and so will `v8`, no difference.
Why is your attitude different now?

**Comment by Ocramius** [comment]
Hmm, I see, indeed I did set that precedent :thinking: 

Ok, should be alright, as long as we mention it in the release notes (and nobody complained last time) :+1:

File: .github/workflows/mutation-tests.yml
Comment (Slamdunk) [inline .github/workflows/mutation-tests.yml]: Needed because of:
1. https://github.com/psr7-sessions/storageless/pull/146/files#diff-5ec4bd5396a261577636d4bc03aea040244331950ebe1a316cb5ba3b18bb89adR167
1. https://github.com/psr7-sessions/storageless/pull/146/files#diff-690b6266871c3ecde938ae69a4fb4c1bf8de4f3ad6dfea595ba1b18c8c05611fR16

File: composer.json
Comment (Slamdunk) [inline composer.json]: Let's wait for a release, in the meantime I've moved the PR forward

File: src/Storageless/Http/SessionMiddleware.php
Comment (Slamdunk) [inline src/Storageless/Http/SessionMiddleware.php:104]: Rsa are not the only asymmetric signing available, there are also EcDSA and EdDSA, the latter the most secure and easy to use.

File: src/Storageless/Http/SessionMiddleware.php
Comment (Slamdunk) [inline src/Storageless/Http/SessionMiddleware.php:121]: For those who construct `SessionMiddleware` by themselves, this will be pretty handy

File: src/Storageless/Http/SessionMiddleware.php
Comment (Slamdunk) [inline src/Storageless/Http/SessionMiddleware.php]: `NOT_BEFORE` claim required by `StrictValidAt` token validator

File: test/StoragelessTest/Http/SessionMiddlewareTest.php
Comment (Slamdunk) [inline test/StoragelessTest/Http/SessionMiddlewareTest.php:595]: HUGE bug introduced [here](https://github.com/psr7-sessions/storageless/pull/119/files#diff-f49a37ecc9fa0480953b1a1534d1cc67755949456a3ad63c07b212e1958359edR556) by myself, I'm so sorry: without a non-numeric key, the `+` operator n arrays overwrite the same-numeric-key entries, thus removing one provider.

Weird that this bug wasn't captured by mutation testing though

File: composer.json
Comment (Slamdunk) [inline composer.json]: Note to self: since this PR will require a new MAJOR, we can bump dflydev/fig-cookies to v3

File: .github/workflows/phpunit.yml
Comment (Ocramius) [inline .github/workflows/phpunit.yml]: This step directly collides with using "locked" dependencies, since `composer remove` modifies `composer.lock`.

As an alternative, let's move `roave/backward-compatibility-check` to a `tools/` directory where `composer.json` and `composer.lock` are configured for it, and we exclude it in `.gitattributes`

File: infection.json.dist
Comment (Ocramius) [inline infection.json.dist:15]: Shouldn't be ignored: alternatively, we can use `roave/infection-static-analysis-plugin`, which should raise mutation score by a lot.

File: src/Storageless/Http/SessionMiddleware.php
Comment (Ocramius) [reply src/Storageless/Http/SessionMiddleware.php]: Should the `$this->clock->now()` result be reused for both `->issuedAt()` and `canOnlyBeUsedAfter()`? Would make sense to me, and it would lead to two identical values (good for compression algos)

File: src/Storageless/Http/SessionMiddleware.php
Comment (Ocramius) [inline src/Storageless/Http/SessionMiddleware.php]: `$this->clock->now()` should be called only once in this method, IMO.

File: test/StoragelessTest/Http/SessionMiddlewareTest.php
Comment (Ocramius) [reply test/StoragelessTest/Http/SessionMiddlewareTest.php:595]: Is the bug from #119 also fixed in current stable? I don't remember seeing a fix :| 

EDIT: solved also in the previous oldstable branch.

**Review by Ocramius — CHANGES_REQUESTED** [review]

File: test/StoragelessTest/Http/SessionMiddlewareTest.php
Comment (Slamdunk) [reply test/StoragelessTest/Http/SessionMiddlewareTest.php:595]: No, would you like a separate PR for stable branch?

Comment (Ocramius) [reply test/StoragelessTest/Http/SessionMiddlewareTest.php:595]: Yes please!  If it's broken, we should fix it :|

File: .github/workflows/phpunit.yml
Comment (Slamdunk) [reply .github/workflows/phpunit.yml]: Done

File: src/Storageless/Http/SessionMiddleware.php
Comment (Slamdunk) [reply src/Storageless/Http/SessionMiddleware.php]: Done

File: infection.json.dist
Comment (Slamdunk) [reply infection.json.dist:15]: So, I've narrowed the ignored code to be specific for the case.
But I don't have enough energy to learn `roave/infection-static-analysis-plugin` in the next dozen of days.
What I can say is that the `assert()` line is only needed to make Psalm happy: the finger should be pointed at `lcobucci/jwt` which doesn't `@psalm` annotate well enough its `Parser` implementation, see

https://github.com/lcobucci/jwt/blob/2f533837091d0b76a89a059e7ed2b2732b2f459e/src/Token/Parser.php#L26-L37

Comment (Ocramius) [reply infection.json.dist:15]: Let's roll with this for now :+1: 

As for `roave/infection-static-analysis-plugin`, it is a drop-in replacement for `vendor/bin/infection` that behaves the exact same way.

I'll raise a new issue separately, or perhaps introduce it myself when I got time to tinker around this again :+1:

Comment (Slamdunk) [reply infection.json.dist:15]: Ok, on my way to the rebase

**Comment by Ocramius** [comment]
I need to branch off `8.0.x` from `7.2.0` before merging this (since `7.3.0` includes #152)

**Comment by Slamdunk** [comment]
I'm sorry I don't understand: if you branch off `8.0` from `7.4` should be fine, as long as I rebase this onto `7.4` before, no?

**Comment by Ocramius** [comment]
If I branch off off `7.4.x`, won't it be a pain for you to rebase? Or would you prefer doing that?

**Comment by Slamdunk** [comment]
I know the codebase now, go ahead with branching of from the latest one (`7.4.x`), no worry, I'm already mid-way the rebase

**Comment by Ocramius** [comment]
@Slamdunk thanks! If you prefer me doing that, poke me tough :+1:

**Review by Ocramius — APPROVED** [review]
LGTM, thanks @Slamdunk!

**Pull request merged**
